sources/ghostdog/src/infiniband.rs

```diff
@@ -0,0 +1,197 @@
+    for device in sys_devices_dirs {
+        let device = device.context(error::InfinibandDeviceSnafu)?;
+        let device_name: OsString = device.file_name();
+        found_devices.push(SysDevice { name: device_name })
+    }
+
+    found_devices.sort_by_key(|device| device.name.clone());
+    Ok(found_devices)
+}
+
+/// Returns true if the SW_MNG_VALUE is found in the Vital Product Data file for given sysfs device
+pub(crate) fn is_device_sw_mng(device: OsString) -> Result<bool> {
+    let vpd_file = Path::new(SYS_CLASS_DIR)
+        .join(SW_MNG_CLASS)
+        .join(device)
+        .join("device")
+        .join("vpd");
+    if vpd_file.exists() {
+        let vpd_file_bytes: &[u8] =
+            &std::fs::read(vpd_file.as_path()).context(error::ReadFileSnafu {
+                path: vpd_file.as_path(),
+            })?;
+        let needle = SW_MNG_VALUE.as_bytes();
+        if vpd_file_bytes
+            .windows(needle.len())
+            .any(|window| window == needle)
+        {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+/// Struct that holds the two important values for an Infiniband Port when checking for
+/// Subnet Management capabilties.
+#[derive(Clone, Debug)]
+pub(crate) struct InfinibandPort {
+    /// Capability Mask is read in as a string such as 0xa751e848
+    pub(crate) capability_mask: CapabilityMask,
+    /// Port GUID is a u64 that is read in as a : formated string but written out in a form such as 0x
+    pub(crate) port_guid: Guid,
+}
+
+impl InfinibandPort {
+    /// is_sm_enabled checks the specific capability bit for Subnet Management means its enabled
+    // The actual bit means `isSMDisabled`, so check if its not zero
+    pub(crate) fn is_sm_enabled(&self) -> bool {
+        let bit_position: u32 = 10;
+        let mask: u32 = 1 << bit_position;
+        if !(&self.capability_mask & mask) != 0 {
+            return true;
+        }
+        false
+    }
+}
+
+// Given a device in sysfs, iterate over the known paths where ports are defined
+// Iterate over the ports and read in the capability mask and Port GUID for each port
+pub(crate) fn find_ports_for_device(device: OsString) -> Result<Vec<InfinibandPort>> {
+    let mut found_ports: Vec<InfinibandPort> = vec![];
+    let ports_path = Path::new(SYS_CLASS_DIR)
+        .join(SW_MNG_CLASS)
+        .join(device)
+        .join("ports");
+
+    let port_directories = fs::read_dir(ports_path).context(error::InfinibandSysDevicesSnafu)?;
+    for port in port_directories {
+        let mut capability_mask: String = "".to_string();
+        let mut first_guid: String = "".to_string();
+
+        let port = port.context(error::InfinibandSysDevicesSnafu)?;
+        let capability_mask_path = port.path().join("cap_mask");
+        if capability_mask_path.exists() {
+            capability_mask = std::fs::read_to_string(capability_mask_path.as_path()).context(
```

Comment:
This will let you declare the `capability_mask` and `first_guid` bindings at the time you assign to them, and avoid the empty string checks at the bottom of the loop:
```
        if !capability_mask_path.exists() {
            continue;
        }
        ...
        if !first_guid_path.exists() {
            continue;
        }
```